### PR TITLE
Adapt inx-mqtt to iota 2.0

### PR DESCRIPTION
**Pull request by muXxer**

### CODE REVIEW

File: components/mqtt/publish.go [hunk 1]
```diff
@@ -43,30 +43,51 @@ func (s *Server) PublishOnTopic(topic string, payload interface{}) {
...
 
-	s.PublishOnTopicIfSubscribed(topic, &milestoneInfoPayload{
-		Index:       ms.Milestone.Index,
-		Time:        ms.Milestone.Timestamp,
-		MilestoneID: ms.MilestoneID.ToHex(),
+	s.PublishRawOnTopicIfSubscribed(topic, rawCommitment)
+}
+
+func (s *Server) PublishCommitmentInfoOnTopic(topic string, id iotago.CommitmentID) {
+	s.PublishOnTopicIfSubscribed(topic, &commitemntInfoPayload{
```
Comment (muXxer) [inline components/mqtt/publish.go:61]: Typo here

File: components/mqtt/publish.go [hunk 2]
```diff
@@ -43,30 +43,51 @@ func (s *Server) PublishOnTopic(topic string, payload interface{}) {
 	s.sendMessageOnTopic(topic, jsonPayload)
 }
 
-/*
-func (s *Server) PublishMilestoneOnTopic(topic string, ms *nodebridge.Milestone) {
-	if ms == nil || ms.Milestone == nil {
+func (s *Server) PublishCommitmentOnTopic(topic string, commitment *iotago.Commitment) {
```
Comment (muXxer) [inline components/mqtt/publish.go:46]: Maybe change to PublishRawCommitmentOnTopic ?

File: components/mqtt/publish.go [hunk 3]
```diff
@@ -43,30 +43,51 @@ func (s *Server) PublishOnTopic(topic string, payload interface{}) {
...
-	block, err := blk.UnwrapBlock(serializer.DeSeriModeNoValidation, nil)
+	block, err := blk.UnwrapBlock(apiForVersion)
 	if err != nil {
 		return
 	}
 
 	s.PublishRawOnTopicIfSubscribed(topicBlocks, blk.GetData())
 
-	switch payload := block.Payload.(type) {
+	basicBlk, isBasicBlk := block.Block.(*iotago.BasicBlock)
```
Comment (muXxer) [inline components/mqtt/publish.go:85]: So we never publish `ValidationBlock`s via INX? Mhh but maybe that could be needed as well :thinking:

File: components/mqtt/publish.go [hunk 4]
```diff
@@ -175,47 +149,29 @@ func (s *Server) PublishBlockMetadata(metadata *inx.BlockMetadata) {
...
+	rawOutputJSON, err := api.JSONEncode(iotaOutput)
 	if err != nil {
 		return nil
 	}
 
-	outputID := output.GetOutputId().Unwrap()
 	rawRawOutputJSON := json.RawMessage(rawOutputJSON)
 
 	return &outputPayload{
-		Metadata: &outputMetadataPayload{
```
Comment (muXxer) [inline components/mqtt/publish.go:196]: So we don't want to pass metadata anymore together with the output?

File: components/mqtt/publish.go [hunk 5]
```diff
@@ -313,17 +269,18 @@ func (s *Server) PublishOnOutputChainTopics(outputID iotago.OutputID, output iot
 	}
 }
 
-func (s *Server) PublishOutput(ctx context.Context, ledgerIndex uint32, output *inx.LedgerOutput, publishOnAllTopics bool) {
-
-	iotaOutput, err := output.UnwrapOutput(serializer.DeSeriModeNoValidation, nil)
+func (s *Server) PublishOutput(ctx context.Context, ledgerIndex iotago.SlotIndex, output *inx.LedgerOutput, publishOnAllTopics bool) {
+	// get api by verson or ledgerIndex?
```
Comment (muXxer) [inline components/mqtt/publish.go:273]: I guess current is fine? Maybe @alexsporn can say something here?

File: components/mqtt/publish.go [hunk 6]
```diff
@@ -350,17 +307,46 @@ func (s *Server) PublishOutput(ctx context.Context, ledgerIndex uint32, output *
...
+	}
 
-	iotaOutput, err := spent.GetOutput().UnwrapOutput(serializer.DeSeriModeNoValidation, nil)
+	// Serialize here instead of using publishOnTopic to avoid double JSON marshaling
+	jsonPayload, err := json.Marshal(response)
+	if err != nil {
+		return
+	}
+
+	if hasOutputMetadataTopicSubscriber {
```
Comment (muXxer) [inline components/mqtt/publish.go:334]: Why do you need to check again? isn't that checked some lines above?

File: components/mqtt/types.go [hunk 7]
```diff
@@ -6,39 +6,27 @@ import (
...
-// milestoneInfoPayload defines the payload of the milestone latest and confirmed topics.
-type milestoneInfoPayload struct {
-	// The index of the milestone.
-	Index uint32 `json:"index"`
-	// The unix time of the milestone payload.
-	Time uint32 `json:"timestamp"`
-	// The ID of the milestone.
-	MilestoneID string `json:"milestoneId"`
+// commitemntInfoPayload defines the payload of the commitment latest and confirmed topics.
+type commitemntInfoPayload struct {
```
Comment (muXxer) [inline components/mqtt/types.go:10]: typo

File: components/mqtt/types.go [hunk 8]
```diff
@@ -47,28 +35,22 @@ type outputMetadataPayload struct {
...
+	// The commitment ID at which this output was booked into the ledger.
+	IncludedCommitmentID string `json:"includedCommitmentId"`
+	// The latest commitment ID of the node.
+	LatestCommitmentID string `json:"latestCommitmentId"`
 }
 
 // outputPayload defines the payload of the output topics.
 type outputPayload struct {
-	// The metadata of the output.
-	Metadata *outputMetadataPayload `json:"metadata"`
```
Comment (muXxer) [inline components/mqtt/types.go:71]: Not sure we should remove that. :thinking:

File: components/mqtt/topics.go [hunk 9]
```diff
@@ -12,9 +12,9 @@ const (
 	parameterCondition     = "{condition}"
 	parameterAddress       = "{address}"
 
-	//topicMilestoneInfoLatest    = "milestone-info/latest"    // milestoneInfoPayload
-	//topicMilestoneInfoConfirmed = "milestone-info/confirmed" // milestoneInfoPayload
-	//topicMilestones             = "milestones"               // iotago.Milestone serialized => []bytes
+	topicCommitmentInfoLatest    = "commitment-info/latest"    // milestoneInfoPayload
+	topicCommitmentInfoFinalized = "commitment-info/finalized" // milestoneInfoPayload
+	topicCommitments             = "commitments"               // iotago.Milestone serialized => []bytes
```
Comment (muXxer) [inline components/mqtt/topics.go:17]: Change the comment.

File: components/mqtt/server.go [hunk 10]
```diff
@@ -438,23 +410,48 @@ func (s *Server) listenToLedgerUpdates(ctx context.Context) error {
...
-func (s *Server) publishLatestMilestoneTopic() {
-	s.LogDebug("publishLatestMilestoneTopic")
-	latest, err := s.NodeBridge.LatestMilestone()
-	if err == nil {
-		s.PublishMilestoneOnTopic(topicMilestoneInfoLatest, latest)
+func (s *Server) publishLatestCommitmentTopic() {
+	s.LogDebug("publishLatestCommitmentTopic")
+	latest, err := s.NodeBridge.LatestCommitment()
+	if err != nil {
+		s.LogError("failed to retrieve latest commitment")
```
Comment (muXxer) [inline components/mqtt/server.go:417]: you should return here if there was an error

File: components/mqtt/server.go [hunk 11]
```diff
@@ -438,23 +410,48 @@ func (s *Server) listenToLedgerUpdates(ctx context.Context) error {
...
-func (s *Server) publishConfirmedMilestoneTopic() {
-	s.LogDebug("publishConfirmedMilestoneTopic")
-	confirmed, err := s.NodeBridge.ConfirmedMilestone()
-	if err == nil {
-		s.PublishMilestoneOnTopic(topicMilestoneInfoConfirmed, confirmed)
+func (s *Server) publishLatestCommitmentInfoTopic() {
+	s.LogDebug("publishLatestCommitmentInfoTopic")
+	latest, err := s.NodeBridge.LatestCommitment()
+	if err != nil {
+		s.LogError("failed to retrieve latest commitment")
```
Comment (muXxer) [inline components/mqtt/server.go:427]: you should return here if there was an error

File: components/mqtt/server.go [hunk 12]
```diff
@@ -438,23 +410,48 @@ func (s *Server) listenToLedgerUpdates(ctx context.Context) error {
...
 	}
+
+	id, err := latest.ID()
+	if err != nil {
+		s.LogError("failed to retrieve latest commitment")
```
Comment (muXxer) [inline components/mqtt/server.go:432]: you should return here if there was an error

File: components/mqtt/server.go [hunk 13]
```diff
@@ -438,23 +410,48 @@ func (s *Server) listenToLedgerUpdates(ctx context.Context) error {
...
+
+	s.LogDebug("publishLatestCommitmentTopic")
+	s.PublishCommitmentOnTopic(topicCommitmentInfoLatest, latest)
+}
+
+func (s *Server) publishFinalizedCommitmentInfoTopic() {
+	s.LogDebug("publishFinalizedCommitmentInfoTopic")
+	finalized, err := s.NodeBridge.LatestFinalizedCommitment()
+	if err != nil {
+		s.LogError("failed to retrieve latest commitment")
```
Comment (muXxer) [inline components/mqtt/server.go:445]: you should return here if there was an error

File: components/mqtt/server.go [hunk 14]
```diff
@@ -438,23 +410,48 @@ func (s *Server) listenToLedgerUpdates(ctx context.Context) error {
...
+	}
+
+	id, err := finalized.ID()
+	if err != nil {
+		s.LogError("failed to retrieve latest finalized commitment")
```
Comment (muXxer) [inline components/mqtt/server.go:450]: you should return here if there was an error

File: components/mqtt/server.go [hunk 15]
```diff
@@ -471,7 +468,16 @@ func (s *Server) fetchAndPublishOutput(ctx context.Context, outputID iotago.Outp
 	if err != nil {
 		return
 	}
-	s.PublishOutput(ctx, resp.GetLedgerIndex(), resp.GetOutput(), false)
+	s.PublishOutput(ctx, iotago.SlotIndex(resp.GetOutput().GetSlotBooked()), resp.GetOutput(), false)
```
Comment (muXxer) [inline components/mqtt/server.go:471]: I think the "slot booked" is the wrong index. It should be the commitment ID of the ledger instead IMO.

File: components/mqtt/server.go [hunk 16]
```diff
@@ -471,7 +468,16 @@ func (s *Server) fetchAndPublishOutput(ctx context.Context, outputID iotago.Outp
...
+}
+
+func (s *Server) fetchAndPublishOutputMetadata(ctx context.Context, outputID iotago.OutputID) {
+	s.LogDebugf("fetchAndPublishOutputMetadata: %s", outputID.ToHex())
+	resp, err := s.NodeBridge.Client().ReadOutputMetadata(ctx, inx.NewOutputId(outputID))
+	if err != nil {
+		return
```
Comment (muXxer) [inline components/mqtt/server.go:478]: Maybe log the error like in the other cases?

File: components/mqtt/server.go [hunk 17]
```diff
@@ -471,7 +468,16 @@ func (s *Server) fetchAndPublishOutput(ctx context.Context, outputID iotago.Outp
 	if err != nil {
 		return
```
Comment (muXxer) [inline components/mqtt/server.go:469]: Maybe also log the error here?

File: components/mqtt/publish.go (hunk 5 above)
Comment (alexsporn) [reply components/mqtt/publish.go:273]: current should ne fine

File: components/mqtt/publish.go (hunk 6 above)
Comment (jkrvivian) [reply components/mqtt/publish.go:334]: Exactly!!

File: components/mqtt/server.go (hunk 15 above)
Comment (jkrvivian) [reply components/mqtt/server.go:471]: This index is not used in the method, I'll remove it.

**Pull request merged**